pandas/tests/test_groupby.py

```diff
@@ -690,6 +691,24 @@ def test_agg_period_index(self):
         grouped = df.groupby(df.index.month)
         list(grouped)
 
+    def test_agg_dict_parameter_cast_result_dtypes(self):
+        # GH 12821
+
+        df = DataFrame(
+            {'class': ['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D'],
+             'time': date_range('1/1/2011', periods=8, freq='H')})
+        df.loc[[0, 1, 2, 5], 'time'] = None
+
+        exp = df.groupby('class').first()
+
+        res = df.groupby('class').agg('first')
+        assert_attr_equal('dtypes', df['time'], res['time'])
```

Comment:
you don't need to ``assert_attr_equal`` the ``assert_frame_equal`` is enough, BUT actually construct it from a frame (and not the result of another operation),  and test the ``.first()`` as well.